### PR TITLE
FindMeetingQuery.java implementation

### PR DESCRIPTION
**Pull request by lattai**
Implemented FindMeetingQuery.java to find a meeting that can be attended by all attendees given the other events happening and meeting info

### CODE REVIEW

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/TimeRange.java [hunk 1]
```diff
@@ -21,166 +21,166 @@
  * providing methods to make ranges easier to work with (e.g. {@code overlaps}).
```
Comment (lattai) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/TimeRange.java:21]: I didn't change anything here, just indentations

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 2]
```diff
@@ -15,9 +15,78 @@
...
+            return avaliableTimes;
+        }
+
+        // Get Collection of unavaliableTimes times
+        Collection<TimeRange> unavaliableTimes = getUnavaliableTimes(events, attendees);
+        // Sort Unavaliable times by start time
+        List<TimeRange> unavTimesList = unavaliableTimes.stream().sorted(TimeRange.ORDER_BY_START).collect(Collectors.toList());
+        // Create Avaliable times
+        avaliableTimes = generateAvaliableTimes(unavTimesList, duration);
+        return avaliableTimes;
```
Comment (lzoebaker) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:46]: Great job on encapsulating logic into clear methods!

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 3]
```diff
@@ -15,9 +15,78 @@
...
+    }
+
+    public Collection<TimeRange> getUnavaliableTimes(Collection<Event> events, Collection<String> attendees) {
+        Collection<TimeRange> badTimes = new ArrayList<>();
+        int i; // Used to ensure only one attendee has to attend another event to make the time unavaliable
```
Comment (lzoebaker) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:51]: I would replace i with a more descriptive variable. Not quite sure what 'i' is.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 4]
```diff
@@ -15,9 +15,78 @@
...
+        for (Event event : events){
+            Set<String> eventAttendees = event.getAttendees();
```
Comment (lzoebaker) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:53]: Could just replace this logic with checking if the two collections are disjoint? https://docs.oracle.com/javase/7/docs/api/java/util/Collections.html#disjoint(java.util.Collection,%20java.util.Collection)

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 5]
```diff
@@ -15,9 +15,78 @@
...
+        return badTimes;
+    }
+
+    public Collection<TimeRange> generateAvaliableTimes(List<TimeRange> unavaliableTimes, long duration) {
+        Collection<TimeRange> avTimes = new ArrayList<>();
+        TimeRange avTime = TimeRange.WHOLE_DAY;
+        for (TimeRange unAvTime : unavaliableTimes){
+            if (unAvTime.overlaps(avTime)) {
+                // Create new Time for av Times
+                TimeRange newTime = TimeRange.fromStartEnd(avTime.start(), unAvTime.start(), false);
```
Comment (lzoebaker) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:71]: Maybe add a comment in front of the boolean, like `/* inclusive of end= */ false`

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 6]
```diff
@@ -15,9 +15,78 @@
...
+            }
+        }
+        // Adding rest of day if there's time left
+        if (avTime.duration() > 0){
+        avTimes.add(avTime);
+        }
+        return avTimes;
+    }
+    
+    public boolean compareDuration(TimeRange newTime, long duration) {
```
Comment (lzoebaker) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:86]: I might change the method name to be a bit more descriptive, like 
`isDurationLessThanTimeRangeDuration()`

**Review by lzoebaker — APPROVED** [review]
Great work Leah! I like your use of helper functions!

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 3 above)
Comment (menynexi) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:51]: +1

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 5 above)
Comment (menynexi) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:71]: +1

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 7]
```diff
@@ -15,9 +15,78 @@
...
+                    i = 1;
+                }
+            }
+        }
+        return badTimes;
+    }
+
+    public Collection<TimeRange> generateAvaliableTimes(List<TimeRange> unavaliableTimes, long duration) {
+        Collection<TimeRange> avTimes = new ArrayList<>();
+        TimeRange avTime = TimeRange.WHOLE_DAY;
```
Comment (menynexi) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:67]: averageTime ?? maybe replace it with a longer but clearer variable name

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/TimeRange.java (hunk 1 above)
Comment (menynexi) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/TimeRange.java:21]: ;) good job

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/TimeRange.java [hunk 8]
```diff
@@ -21,166 +21,166 @@
...
+    public int hashCode() {
+        return Long.hashCode(start) ^ Long.hashCode(duration);
     }
 
-    // If the other range has no duration, then we must treat it like a point in time rather than a
-    // range.
-    if (other.duration <= 0) {
-      return contains(this, other.start);
+    @Override
+    public String toString() {
```
Comment (menynexi) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/TimeRange.java:134]: might want to change the name to formattedString() instead?

**Review by menynexi — APPROVED** [review]

**Comment by DirtGrubDylan** [comment]
Didn't want me to review it? :(

**Pull request merged**